Publish dotnet/runtime using standard Arcade Publish tooling

jkoritzinsky opened the pull request:
Publish the assets from dotnet/runtime using the standard Arcade publish tooling. This tooling doesn't allow any duplicate packages, so change our publishing logic to the following:

- In a given build, publish all packages that were produced in the current build that contain that job's target RID.
- For some target RIDs, also publish an additional subset of packages (specified in `Publishing.props`).
- For some jobs (specified in runtime-official), publish all packages produced.
- In the VMR, mark all "packages not to be published" as "Vertical" visibility so they're still available upstack.
- Outside the VMR, skip publishing all "packages not to be published" packages.

~~Blocked on MicroBuild fixes for Linux and Mac: https://github.com/dotnet/source-build/issues/4793~~ As per https://github.com/dotnet/runtime/pull/111934#issuecomment-2705190197, we are no longer blocked on this.

Verified in https://dev.azure.com/dnceng/internal/_build/results?buildId=2629649&view=results that all of the same assets that we expect to produce in a runtime official build are produced on this branch.

Fixes https://github.com/dotnet/runtime/issues/102672
Fixes https://github.com/dotnet/runtime/issues/1719

jkoritzinsky committed:
Remove prepare-signed-artifacts and publish from each leg directly.

jkoritzinsky committed:
Publish in a separate stage to make dependency tracking easy

jkoritzinsky committed:
Set _SignType so we can pass -sign

jkoritzinsky committed:
Remove old validation target

jkoritzinsky committed:
!drop remove -sign so we can figure out publishing first

jkoritzinsky committed:
Remove InitialTargets

jkoritzinsky committed:
Only generate product version files once. Fix asset manifest extension. Don't copy workloads packages as they're already in the right format now.

jkoritzinsky committed:
Completely restructure how we publish to be more explicit that specific jobs publish specific packages

jkoritzinsky committed:
Move "non-RID-specific" special packages logic into Publishing.props

jkoritzinsky committed:
Make sure we build the list of skipped packages after we add the RID-specific packages

jkoritzinsky committed:
Arcade now handles TargetOS/TargetArchitecture for asset manifest names. Use that when possible for unique names and produce our own names when necessary.

jkoritzinsky committed:
Extract handling hostOS and hostArch into another props file

jkoritzinsky committed:
Fix import and adjust condition for SB

jkoritzinsky committed:
Fix asset manifest name uniqueness to ensue we don't overwrite asset manifests.

jkoritzinsky committed:
Set AssetManifestFilePath again

jkoritzinsky committed:
Fix PublishAllPackages condition and make sure to publish the VS redist packages

jkoritzinsky committed:
Explicitly look up packages as we add zips and other blob artifacts to the Artifacts items in Signing.props

jkoritzinsky committed:
Search in the right folder.

jkoritzinsky committed:
Look up zips in VSSetup for workloads

jkoritzinsky committed:
Add missing semicolon

jkoritzinsky committed:
Include VSSetup artifacts separately

jkoritzinsky committed:
Re-enable signing (depends on Microbuild signing on Linux/Mac)

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into standard-arcade-publish

ViktorHofer commented on eng/pipelines/runtime-official.yml:
```diff
@@ -394,13 +327,9 @@ extends:
           jobParameters:
             templatePath: 'templates-official'
-            buildArgs: -s tools+libs -pack -c $(_BuildConfig) /p:TestAssemblies=false /p:TestPackages=true
+            buildArgs: -s tools+libs -restore -build -pack -sign -publish -c $(_BuildConfig) /p:TestAssemblies=false /p:TestPackages=true /p:PublishAllPackages=true /p:AssetManifestFileName=LibrariesPackages.xml
```
Can we just use the job name for the manifest file in general?

ViktorHofer commented on eng/Signing.props:
```diff
@@ -45,30 +41,26 @@
-              SkipPublish="true" />
+    <Artifact Include="$(ArtifactsDir)VSSetup\**\*.zip"
+              Condition="Exists('$(ArtifactsDir)VSSetup')"
+              IsShipping="false">
+      <ChecksumPath>%(FullPath).sha512</ChecksumPath>
+      <SubBlobFolder>workloads/</SubBlobFolder>
+    </Artifact>
```
nit:

```suggestion
              IsShipping="false"
              ChecksumPath="%(FullPath).sha512"
              SubBlobFolder="workloads/" />
```

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -1,11 +1,15 @@
-    <!-- Set PlatformName to TargetArchitecture to create unique build manifest files. -->
-    <PlatformName Condition="'$(TargetArchitecture)' != ''">$(TargetArchitecture)</PlatformName>
+    <!-- Except for jobs that set a specific asset manifest file name as a global property, set the manifest name based on the package RID. -->
+    <AssetManifestFileName>$(PackageRID).xml</AssetManifestFileName>
```
As mentioned in another comment, consider using the AGENT_JOBNAME env var here. This should into Arcade ideally:

```suggestion
    <AssetManifestFileName Condition="'$(AGENT_JOBNAME)' != ''">$(AGENT_JOBNAME)</AssetManifestFileName>
```

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -169,14 +120,13 @@
   <PropertyGroup>
-    <_ShouldGenerateProductVersionFiles Condition="'$(DotNetBuildRepo)' == 'true' and '$(DotNetBuildOrchestrator)' == 'true'">true</_ShouldGenerateProductVersionFiles>
-    <_ShouldGenerateProductVersionFiles Condition="'$(DotNetBuildRepo)' != 'true' and '$(DotNetBuildOrchestrator)' != 'true'">true</_ShouldGenerateProductVersionFiles>
+    <ShouldGenerateProductVersionFiles Condition="'$(DotNetBuildOrchestrator)' == 'true'">true</ShouldGenerateProductVersionFiles>
```
IIRC we only need that file produced once, i.e. in win-x64.

jkoritzinsky replied:
I have the build set up to drive it in the runtime-official pipeline (producing it in the workloads job as that's a little easier).

In the VMR, I'm pretty sure that the productVersion.txt is used downstream (I think aspnetcore might reference it? I'm not 100% positive where I saw it).

mmitche replied:
> In the VMR, I'm pretty sure that the productVersion.txt is used downstream (I think aspnetcore might reference it? I'm not 100% positive where I saw it).

I don't think it does? THere is some logic I can see that looks for the files in arcade, and produces one in aspnecore, but no uses I can see in aspnetcore.

ViktorHofer replied:
Yeah I would surprised if that file is read by aspnetcore in the VMR build.

jkoritzinsky replied:
In that case, I'll adjust this to run in win-x64 BuildPass1 in the VMR.

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -31,125 +35,72 @@
-    <Artifact Include="@(WorkloadsVSInsertionFile)"
-              SubBlobFolder="workloads/"
-              IsShipping="$([System.String]::new('%(Identity)').Contains('\Shipping\'))" />
+    <Artifact Update="@(PackagesToSkipFromThisJob)" SkipPublish="true" Condition="'$(DotNetBuildOrchestrator)' != 'true'" />
```
I'm not sure I understand why you set the `SkipPublish` metadata here instead of just removing them items?

ViktorHofer replied:
I would prefer to use standard msbuild functionality here (item `Remove` action) than relying on custom metadata. Not just because I question the use-case for the `SkipPublish` metadata in general when we stop importing Sign.props in Publish.proj (hehe, that's what you get when you introduce new extension points in API) but because I find the control harder to follow with this.

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -31,125 +35,72 @@
+    <Artifact Update="@(PackagesToSkipFromThisJob)" Visibility="Vertical" Condition="'$(DotNetBuildOrchestrator)' == 'true'" />
   </ItemGroup>
 
+  <!-- Include any additional packages specified by the job here. -->
+  <Target Name="IncludeAdditionalSpecificPackages" BeforeTargets="BeforePublish" Inputs="@(AdditionalSpecificPackagesToPublish)" Outputs="@(AdditionalSpecificPackagesToPublish->'%(Identity).batching')">
+    <ItemGroup>
+      <Artifact Condition="$([System.String]::new('%(Filename)').StartsWith('%(AdditionalSpecificPackagesToPublish.Identity)'))"
+                SkipPublish="false"
+                Visibility="$(DefaultArtifactVisibility)" />
+    </ItemGroup>
+  </Target>
```
I don't see where `AdditionalSpecificPackagesToPublish` is defined.

jkoritzinsky committed:
PR feedback

jkoritzinsky committed:
Specify Apple file signing info on macOS and simplify the macOS jobs in the official build.

mmitche commented on eng/Signing.props:
```diff
@@ -14,6 +13,9 @@
     <FileSignInfo Include="mscordbi.dll" CertificateName="None" />
 
+    <!-- On MacOS, we need to sign a number of our executables with the Mac developer cert with hardening enabled. -->
+    <FileSignInfo Condition="'$(TargetsOSX)' == 'true'" Include="dotnet;apphost;corerun;createdump" CertificateName="MacDeveloperHarden" />
```
Will these files have the appropriate entitlements before signing?

jkoritzinsky replied:
Yes, we entitle and ad-hoc sign them during the initial CMake build. As long as this preserves entitlements (which as far as I can tell from docs does), we should be good.

mmitche replied:
Yes, it will.

mmitche reviewed (approved):
This is a fantastic change

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -31,125 +32,64 @@
+    <Artifact Update="@(PackagesToSkipFromThisJob)" Visibility="Vertical" Condition="'$(DotNetBuildOrchestrator)' == 'true'" />
   </ItemGroup>
 
+
```
nit remove the additional empty line:

```suggestion
```

jkoritzinsky committed:
In a VMR-style build, produce the product version file in either the win-x64 leg (for UB) or in all legs (for SB)

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -168,15 +107,15 @@
-  <PropertyGroup>
-    <_ShouldGenerateProductVersionFiles Condition="'$(DotNetBuildRepo)' == 'true' and '$(DotNetBuildOrchestrator)' == 'true'">true</_ShouldGenerateProductVersionFiles>
-    <_ShouldGenerateProductVersionFiles Condition="'$(DotNetBuildRepo)' != 'true' and '$(DotNetBuildOrchestrator)' != 'true'">true</_ShouldGenerateProductVersionFiles>
+  <PropertyGroup Condition="'$(DotNetBuildOrchestrator)' == 'true'">
```
Don't we need to produce the version files in the current msft builds as well?

jkoritzinsky replied:
They're produced by the workloads job (set in runtime-official.yml)

ViktorHofer replied:
They were previously also produced when building locally. I assume this isn't the case anymore?

jkoritzinsky replied:
Yeah they won't be produced locally now unless you request them

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -1,11 +1,12 @@
     <AutoGenerateSymbolPackages>false</AutoGenerateSymbolPackages>
-    <!-- Set PlatformName to TargetArchitecture to create unique build manifest files. -->
-    <PlatformName Condition="'$(TargetArchitecture)' != ''">$(TargetArchitecture)</PlatformName>
+    <!-- Except for jobs that set a specific asset manifest file name as a global property, set the manifest name based on the package RID. -->
+    <AssetManifestFileName Condition="'$(AGENT_JOBNAME)' != ''">$(AGENT_JOBNAME)</AssetManifestFileName>
+    <!-- Reset the file path now that we may have changed the name. -->
+    <AssetManifestFilePath>$(ArtifactsLogDir)AssetManifest\$(AssetManifestFileName)</AssetManifestFilePath>
```
You can revert this after https://github.com/dotnet/runtime/pull/112013 is merged:

```suggestion
```

jkoritzinsky commented:
I've verified that signing works for the Mac assets as expected, so we're good to go on that front (once the parallelization issues are fixed)

jkoritzinsky committed:
Change where we set PublishAllPackages to make things automatically work when this change flows to the VMR.

jkoritzinsky committed:
Merge branch 'main' into standard-arcade-publish

jkoritzinsky committed:
Update eng/Publishing.props

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

ViktorHofer commented on eng/DotNetBuild.props:
```diff
@@ -103,6 +104,14 @@
+        In the future, some VMR jobs may not publish all packages.
+        SourceBuild will always need to publish everything.
+      -->
+      <InnerBuildArgs Condition="'$(ShortStack)' != 'true'">$(InnerBuildArgs) /p:PublishAllPackages=true</InnerBuildArgs>
```
How is `PublishAllPackages` different from `EnableDefaultArtifacts`?

jkoritzinsky replied:
I was trying to keep EnableDefaultArtifacts set per your feedback on a prior PR. We could change this to use that instead if you'd prefer.

ViktorHofer replied:
I don't think that I meant that it should always be used in every job, back in that other PR. The default experience should be to have it default to true (i.e. locally or on the designated primary vertical -> `win-64` which produces all the rid agnostic packages and passes the `-pack` switch in).

Could you please share a commit that shows how things would change when using `EnableDefaultArtifacts` as the control flow property?

jkoritzinsky replied:
I've pushed an update to use `EnableDefaultArtifacts` here instead.

jkoritzinsky committed:
Change the control property to be EnableDefaultArtifacts and Include instead of Update for these scenarios.

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -1,154 +1,108 @@
-    <TransportPackagesToPublishFile
-      Include="$(DownloadDirectory)**\*Transport*.nupkg"
-      Exclude="@(RuntimeNupkgFile);@(RidAgnosticNupkgToPublishFile);@(DownloadedSymbolNupkgFile)" />
+    <EnableDefaultArtifacts>false</EnableDefaultArtifacts>
```
Can we set the default that is currently set in DotNetBuild.props directly here?

```suggestion
    <EnableDefaultArtifacts Condition="'$(DotNetBuildOrchestrator)' != 'true'">false</EnableDefaultArtifacts>
```

... and pass false in, from DotNetBuild.props when doing a short stack build.

ViktorHofer commented on eng/Publishing.props:
```diff
@@ -1,154 +1,108 @@
-    <TransportPackagesToPublishFile
-      Include="$(DownloadDirectory)**\*Transport*.nupkg"
-      Exclude="@(RuntimeNupkgFile);@(RidAgnosticNupkgToPublishFile);@(DownloadedSymbolNupkgFile)" />
+    <EnableDefaultArtifacts>false</EnableDefaultArtifacts>
```
Should this default to true when building locally?

jkoritzinsky replied:
I don't like having a "CI vs local build" difference here at that level. I'd rather have our "out of VMR vs in VMR" builds be consistent whether it's a local build or not. Makes debugging issues easier.

jkoritzinsky committed:
Avoid global-property confusion within the VMR scenarios

jkoritzinsky committed:
Merge branch 'main' of https://github.com/dotnet/runtime into standard-arcade-publish

jkoritzinsky committed:
Collapse more jobs together in the YAML

jkoritzinsky committed:
Remove unused items

jkoritzinsky committed:
Move logic into Signing.props.

This solves problems around specifying EnableDefaultArtifacts=false in some VMR legs

jkoritzinsky committed:
Only publish the LLVM runtime packs from LLVM legs

jkoritzinsky committed:
Add the package artifacts to the right item group

jkoritzinsky committed:
Fix package name format

jkoritzinsky committed:
Fix handling host-targeting assets as NETCoreSdkRuntimeIdentifier is not available.

jkoritzinsky committed:
Item updates for default artifacts must happen in Publishing.props

jkoritzinsky committed:
Merge branch 'main' into standard-arcade-publish

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into standard-arcade-publish

jkoritzinsky committed:
Use Arcade's RID-specific publish feature in dotnet/runtime

jkoritzinsky committed:
Fix dependencies

jkoritzinsky committed:
Update host asset publishing

jkoritzinsky committed:
Set TargetRid

jkoritzinsky committed:
Adjust for Azure Linux package names and use artifact kinds

jkoritzinsky committed:
Merge branch 'main' of github.com:dotnet/runtime into standard-arcade-publish

jkoritzinsky committed:
Restore a final join job for signing so we can merge this in and stop having to update the patch.

jkoritzinsky committed:
Remove unused parameter

jkoritzinsky committed:
The build stage shouldn't be a list

jkoritzinsky committed:
Add variable for artifact name

jkoritzinsky committed:
Try another variant

jkoritzinsky committed:
Try to simplify

jkoritzinsky committed:
Another try

jkoritzinsky committed:
Add the target path

jkoritzinsky committed:
Define the artifact name like how we define the build job names for simplicity

jkoritzinsky committed:
Make some items a list

jkoritzinsky committed:
Rewrite how we insert the "other" stages

jkoritzinsky committed:
Remove duplicate pool

jkoritzinsky committed:
Fix stage name

jkoritzinsky committed:
Fix conditions

jkoritzinsky committed:
Fix expected pipeline artifact names

jkoritzinsky committed:
Always put a templateContext into global-build-job even if one isn't provided

jkoritzinsky committed:
Fix typo and adjust publishing to handle PDBs

jkoritzinsky committed:
Set blob artifact kind

jkoritzinsky committed:
Fix item name for AzureLinux RPM signing

jkoritzinsky committed:
Put the VSSetup blobs in the VSSetup folder and adjust the relative blob paths to make that work

jkoritzinsky committed:
Fix crossdac and workloads downloads

jkoritzinsky committed:
inputs happen before clone. Don't download into the sources directory

jkoritzinsky committed:
Fix paths

jkoritzinsky committed:
Skip extracting packages that don't match so we can see the whole list.

jkoritzinsky committed:
Add arm32 downloads for crossdac

jkoritzinsky committed:
Publish crossos dac packages

jkoritzinsky committed:
Don't use RID-specific artifacts for the cross-dac build

jkoritzinsky committed:
Fix source-build and deps

jkoritzinsky committed:
Fix join-job signing

jkoritzinsky committed:
Add back split macos job to the official build. These files need to be signed coming out of runtime and otherwise we'd miss signing them

jkoritzinsky committed:
Merge branch 'standard-arcade-publish' of https://dev.azure.com/dnceng/internal/_git/dotnet-runtime into standard-arcade-publish-with-join

jkoritzinsky committed:
Remove duplicate jobs

jkoritzinsky committed:
Adjust commands for macos split job

jkoritzinsky commented:
I've updated this PR to provide a mechanism for us to merge it today and get rid of the long-running patch.

Here's my basic approach:

- Instead of publishing to AzDO, each runtime vertical will instead publish to disk with a layout like the `artifacts` directory.
- Runtime verticals won't sign in the runtime official build
- We bring back the final join job. This job downloads all of the artifacts from each runtime vertical into its `artifacts` directory and then does a standard arcade sign/publish with the same signing rules as exist in main.
- We still get rid of IntermediateArtifacts and instead use separate pipeline artifacts for each vertical.

I have a passing official build from this PR here: https://dev.azure.com/dnceng/internal/_build/results?buildId=2657037&view=results

jkoritzinsky committed:
Use the right property for the signing items base directory

jkoritzinsky commented:
New official build for a314fc3 at https://dev.azure.com/dnceng/internal/_build/results?buildId=2657219&view=results

ViktorHofer commented on eng/DotNetBuild.props:
```diff
@@ -38,6 +38,19 @@
+    Allow the VMR orchestrator to control whether or not to build rid-specific artifacts,
+    but provide defaults until the VMR orchestrator provides controls in all scenarios.
+  -->
+  <PropertyGroup Condition="'$(EnableDefaultRidSpecificArtifacts)' == ''">
+    <!-- Source-build always needs all artifacts to be published. -->
+    <EnableDefaultRidSpecificArtifacts Condition="'$(DotNetBuildSourceOnly)' != ''">false</EnableDefaultRidSpecificArtifacts>
+    <!-- Short-stack builds should always only publish RID-specific artifacts. -->
+    <EnableDefaultRidSpecificArtifacts Condition="'$(ShortStack)' == 'true'">true</EnableDefaultRidSpecificArtifacts>
+    <!-- If no override is specified, don't use RID-specific publishing. Instead, publish everything. -->
+    <EnableDefaultRidSpecificArtifacts Condition="'$(EnableDefaultRidSpecificArtifacts)' == ''">false</EnableDefaultRidSpecificArtifacts>
+  </PropertyGroup>
```
These settings make sense to me but wouldn't they better fit into the VMR orchestrator as they aren't specific to runtime?

jkoritzinsky replied:
Yes, they do fit better in the VMR. However, we need to put them here first so they don't break the VMR until I change the VMR to pass these options down (also in PR)

jkoritzinsky replied:
Once the VMR controls are in, I'll come back and remove these.

ViktorHofer replied:
Consider filing a tracking issue and linking to it. Also, so that others can follow along and understand the potential for improvement in these code paths.

ViktorHofer commented on eng/Signing.props:
```diff
@@ -43,41 +73,103 @@
+                  Kind="Package" />
+      </ItemGroup>
+    </When>
+    <Otherwise>
+    </Otherwise>
```
```suggestion
```

jkoritzinsky committed:
Update Signing.props

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

jkoritzinsky committed:
Remove duplicate signing inputs

jkoritzinsky commented:
/ba-g timeouts unrelated
